### ISSUE TITLE
VirtualSchedulerTest.Virtual_ThreadSafety() fails sometimes on my machine.

### ISSUE THREAD

**Issue by quinmars** [dotnet/reactive#785]
I run all tests 10 times and the test failed 3 times. Interestingly it did not fail when I ran the test 20 times alone.

Output:
```
Test Name:	ReactiveTests.Tests.VirtualSchedulerTest.Virtual_ThreadSafety
Test FullName:	ReactiveTests.Tests.VirtualSchedulerTest.Virtual_ThreadSafety
Test Source:	C:\Users\peter\Documents\src\Rx.NET\Rx.NET\Source\tests\Tests.System.Reactive\Tests\Concurrency\VirtualSchedulerTest.cs : line 133
Test Outcome:	Failed
Test Duration:	0:00:02,757

Result StackTrace:	
bei Xunit.Assert.True(Nullable`1 condition, String userMessage)
   bei ReactiveTests.Tests.VirtualSchedulerTest.Virtual_ThreadSafety()
Result Message:	
Virtual time 00:00:11, exception System.Exception: Should have thrown!
   bei ReactiveTests.Tests.VirtualSchedulerTest.Virtual_ThreadSafety()
Expected: True
Actual:   False

```

My CPU (tooked from BenchmarkFotNet):

``` ini
BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i5-5200U CPU 2.20GHz (Broadwell), 1 CPU, 4 logical and 2 physical cores
```

**Comment by akarnokd** [comment]
I think an earlier version of the did not throw but. Could you verify it?

**Comment by akarnokd** [comment]
The previous version did not fail if the `AdvanceBy` didn't throw:

https://github.com/dotnet/reactive/blob/rxnet-v4.0.0/Rx.NET/Source/tests/Tests.System.Reactive/Tests/Concurrency/VirtualSchedulerTest.cs#L153

**Comment by quinmars** [comment]
So, if I understand it right. On my machine with only four cores the subscribing thread does not wackup fast enough after the `Task.Delay(10).Wait()`. So we reach the virtual 11 seconds before the subscription actually happens. That's not what we want to test, but it isn't either wrong.

I propose to change the test to:

```csharp
        public void Virtual_ThreadSafety()
        {
            for (var i = 0; i < 10; i++)
            {
                var scheduler = new TestScheduler();
                var seq = Observable.Never<string>();
                var disposable = default(IDisposable);

                var sync = 2;

                Task.Run(() =>
                {
                    if (Interlocked.Decrement(ref sync) != 0)
                    {
                        while (Volatile.Read(ref sync) != 0)
                        {
                            ;
                        }
                    }

                    Task.Delay(10).Wait();

                    disposable = seq.Timeout(TimeSpan.FromSeconds(5), scheduler).Subscribe(s => { });
                });

                var watch = scheduler.StartStopwatch();
                try
                {
                    if (Interlocked.Decrement(ref sync) != 0)
                    {
                        while (Volatile.Read(ref sync) != 0)
                        {
                            ;
                        }
                    }

                    var d = default(IDisposable);
                    while (watch.Elapsed < TimeSpan.FromSeconds(11))
                    {
                        d = Volatile.Read(ref disposable);
                        scheduler.AdvanceBy(50);
                    }

                    if (d != null)
                    {
                        throw new Exception("Should have thrown!");
                    }
                }
                catch (TimeoutException)
                {
                }
                catch (Exception ex)
                {
                    Assert.True(false, string.Format("Virtual time {0}, exception {1}", watch.Elapsed, ex));
                }
                disposable?.Dispose();
            }
```

So only if the subscription actually finished and the timeout was not thrown, the test will fail. There is still a small window for a false negative, but that's better than a false positive.

**Comment by akarnokd** [comment]
If you increase the virtual time limit to 100s: `while (watch.Elapsed < TimeSpan.FromSeconds(100))`, the likelihood of failure decreases. The original test run extremely slowly with dotCover.

**Comment by quinmars** [comment]
But that way it does not slow down dotCover. Wouldn't that be better? There is still the possiblity to trigger potential thread safety bugs.

**Comment by akarnokd** [comment]
The `Task.Run`'s delay is the dominant factor.

**Comment by quinmars** [comment]
I see. Should I still keep the check, just for the case that the increased likelihood is not enough?

**Comment by akarnokd** [comment]
Since there was no forced crash in the original, I guess we can have the `if (d != null)` check.

**Comment by quinmars** [comment]
This can be closed since #786 is closed.
